data_model/src/query/builder.rs

```diff
@@ -0,0 +1,304 @@
+
+    /// Sort the results according to the specified sorting.
+    pub fn with_sorting(self, sorting: Sorting) -> Self {
+        Self { sorting, ..self }
+    }
+
+    /// Only return part of the results specified by the pagination.
+    pub fn with_pagination(self, pagination: Pagination) -> Self {
+        Self { pagination, ..self }
+    }
+
+    /// Change the batch size of the iterable query.
+    ///
+    /// Larger batch sizes reduce the number of round-trips to iroha peer, but require more memory.
+    pub fn with_fetch_size(self, fetch_size: FetchSize) -> Self {
+        Self { fetch_size, ..self }
+    }
+}
+
+impl<E, Q, P> IterableQueryBuilder<'_, E, Q, P>
+where
+    E: QueryExecutor,
+    Q: IterableQuery,
+    Q::Item: HasPredicateBox<PredicateBoxType = P>,
+    IterableQueryBox: From<IterableQueryWithFilterFor<Q>>,
+    Vec<Q::Item>: TryFrom<IterableQueryOutputBatchBox>,
+    <Vec<Q::Item> as TryFrom<IterableQueryOutputBatchBox>>::Error: core::fmt::Debug,
+{
+    /// Execute the query, returning an iterator over its results.
+    pub fn execute(self) -> Result<IterableQueryIterator<E, Q::Item>, E::Error> {
+        let with_filter = IterableQueryWithFilter::new(self.query, self.filter);
+        let boxed: IterableQueryBox = with_filter.into();
+
+        let query = IterableQueryWithParams {
+            query: boxed,
+            params: IterableQueryParams {
+                pagination: self.pagination,
+                sorting: self.sorting,
+                fetch_size: self.fetch_size,
+            },
+        };
+
+        let (first_batch, continue_cursor) = self.query_executor.start_iterable_query(query)?;
+
+        let iterator = IterableQueryIterator::<E, Q::Item>::new(first_batch, continue_cursor)
+            .expect(
+                "iroha returned unexpected type in iterable query. Is there a schema mismatch?",
+            );
+
+        Ok(iterator)
+    }
+
+    /// Execute the query, returning all the results collected into a vector.
+    pub fn execute_all(self) -> Result<Vec<Q::Item>, E::Error> {
+        self.execute()?.collect::<Result<Vec<_>, _>>()
+    }
+
+    /// Execute the query, constraining the number of results to zero or one. Errors if more than one result is returned.
+    pub fn execute_single_opt(self) -> Result<Option<Q::Item>, E::SingleError> {
+        let mut iter = self.execute()?;
+        let first = iter.next().transpose()?;
+        let second = iter.next().transpose()?;
+
+        match (first, second) {
+            (None, None) => Ok(None),
+            (Some(result), None) => Ok(Some(result)),
+            (Some(_), Some(_)) => Err(SingleQueryError::ExpectedOneOrZeroGotMany.into()),
+            (None, Some(_)) => {
+                unreachable!()
+            }
+        }
+    }
+
+    /// Execute the query, constraining the number of results to exactly one. Errors if zero or more than one result is returned.
+    pub fn execute_single(self) -> Result<Q::Item, E::SingleError> {
```

Comment:
do we need to support this special case? seems to me a client should deal with it themselves

Reply:
also, it's not necessary to check that 2nd result doesn't exist. That's an invariant of the blockchain. A user will execute a query and just take the 1st result. They will not be checking if there is more results. So I don't think method is needed in the API

Reply:
> do we need to support this special case? seems to me a client should deal with it themselves

This _is_ a way for the client to deal with it themselves, just reducing boilerplate. The check is fully implemented on the client side.

> it's not necessary to check that 2nd result doesn't exist. That's an invariant of the blockchain.

Just so that we are on the same page: this methods executes an _iterable_ query (which can return different amount of results depending on the parameters). If iroha returns more than one result it may as well be an ill-specified query, and so we should notify the user that their intention is not met. This also has zero additional cost of network traffic in common case of batch size not being 1, so I don't see why shouldn't we check this.

Reply:
> If iroha returns more than one result it may as well be an ill-specified query, and so we should notify the user that their intention is not met

ok, I'm getting your point. But why is this a recoverable error? This indicates a bug and there is no reasonable way to recover from it

Reply:
let's keep it